Sources/SwiftProcessing/Core/Sketch.swift

```diff
@@ -197,11 +198,11 @@ import SceneKit
         initTouch()
         initNotifications()
         sketchDelegate = self as? SketchDelegate
-        createCanvas(0, 0, UIScreen.main.bounds.width, UIScreen.main.bounds.height)
+        createCanvas(0.0, 0.0, UIScreen.main.bounds.width, UIScreen.main.bounds.height)
```

Comment:
are the decimals still needed or could 0 work once again due to the generics?

Reply:
The thing with generics is that you can put any number type in, but all of the arguments need to be the same type. I don't believe there's any way around this behavior, but I believe the benefits outweigh the drawbacks. The benefit is that we don't have to talk about API-specific types in such an explicit way like when the library used CGFloats in its user-facing API.

The behavior you can expect is that  background(0) and background(0.0) are equivalent and will run. When you have more than one argument, that's when you need to be conscious of type. So background(0,0,0) will work, but background(0, 123.3, 0) will throw an error.

Reply:
Hey, I just stumbled across this project today. I hope you don't mind me butting in, but there is a way to allow each parameter to have a different type by making each parameter type a separate generic. For example:

```swift
open func createCanvas<X: Numeric, Y: Numeric, W: Numeric, H: Numeric>(_ x: X, _ y: Y, _ width: W, _ height: H) { // ...
```

It makes the function declaration a bit unwieldy but would give you the calling behavior you're looking for.

Reply:
> It makes the function declaration a bit unwieldy but would give you the calling behavior you're looking for.

Thanks for suggesting this. It's a nice solution.

The only issue I can see is that it may make the source code more difficult to read. We have three target audiences here that I like to think about: API users, API contributors, and learners who like to see how things work. It would be nice if it didn't confuse people looking at the source code.

Is this is a common practice in Swift development?

I'm relatively new to all this, so I'm open to it if @jjkaufman likes the idea.

Reply:
thanks Jake! It makes sense to me. It does expand out the function declaration, but only to do exactly what we want. Each input can be of their own unique Numeric type.

The introduction of generics definitely raises the bar on Swift expertise required to understand the packages source code, but in this case its a net win due to the UX improvement for the usage of the library when making sketches

Reply:
Having spent some time implementing this, it does break default nil initializers because the type can't be inferred for nil. The Vector class and image function are two examples.

Vector class:
```
    public init<X: Numeric, Y: Numeric, Z: Numeric>(_ x: X, _ y: Y, _ z: Z? = nil) {
        self.x = x.convert()
        self.y = y.convert()
        self.z = z?.convert()
    }
```

Image function:
```
 func image<X: Numeric, Y: Numeric, W: Numeric, H: Numeric>(_ image: Image, _ x: X, _ y: Y, _ width: W? = nil, _ height: H? = nil) {
        var cg_x, cg_y: CGFloat
        cg_x = x.convert()
        cg_y = y.convert()
        
        let w:CGFloat = width == nil ? CGFloat(image.width) : width!.convert()
        let h:CGFloat = height == nil ? CGFloat(image.height) : height!.convert()
        
        image.width = Double(w)
        image.height = Double(h)
        
        image.frame(CGFloat(deltaTime), CGFloat(frameCount)).draw(in: CGRect(x: cg_x, y: cg_y, width: w, height: h), blendMode: image.blendMode, alpha: CGFloat(image.alpha))
    }
```

Unless I can find another way to do this, this will require creating a new initializer that removes the nil defaults for any classes that have nil default values and a new overload for functions.

Reply:
Update: New initializers and overloads work fine and was the solution.